src/envoy/http/authn/authn_utils_test.cc

```diff
@@ -0,0 +1,71 @@
+/* Copyright 2018 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "src/envoy/http/authn/authn_utils.h"
+#include "src/envoy/http/authn/authenticator_base.h"
+#include "src/envoy/http/authn/test_utils.h"
+#include "test/test_common/utility.h"
+
+using google::protobuf::util::MessageDifferencer;
+using istio::authn::JwtPayload;
+
+namespace iaapi = istio::authentication::v1alpha1;
+
+namespace Envoy {
+namespace Http {
+namespace Istio {
+namespace AuthN {
+namespace {
+
+const std::string kSecIstioAuthUserInfoHeaderKey = "sec-istio-auth-userinfo";
+const std::string kSecIstioAuthUserinfoHeaderValue =
+    "eyJpc3MiOiI2Mjg2NDU3NDE4ODEtbm9hYml1MjNmNWE4bThvdmQ4dWN2Njk4bGo3OH"
+    "Z2MGxAZGV2ZWxvcGVyLmdzZXJ2aWNlYWNjb3VudC5jb20iLCJzdWIiOiI2Mjg2NDU3"
+    "NDE4ODEtbm9hYml1MjNmNWE4bThvdmQ4dWN2Njk4bGo3OHZ2MGxAZGV2ZWxvcGVyLm"
+    "dzZXJ2aWNlYWNjb3VudC5jb20iLCJhdWQiOiJib29rc3RvcmUtZXNwLWVjaG8uY2xv"
+    "dWRlbmRwb2ludHNhcGlzLmNvbSIsImlhdCI6MTUxMjc1NDIwNSwiZXhwIjo1MTEyNz"
+    "U0MjA1fQ==";
+
+TEST(AuthnUtilsTest, GetJwtPayloadFromHeaderTest) {
```

Comment:
It would be better to test the JsonIteratorCallback directly, with more thorough input cases (claims with string, array, and other primitive types etc)

Reply:
JsonIteratorCallback() is in an anonymous namespace, which is inaccessible from authn_utils_test.cc. Moreover, the const Envoy::Json::Object& parameter in JsonIteratorCallback(...const Envoy::Json::Object&...) is of type Field, which is also in an anonymous namespace and is only creatable in the original cc file.

Reply:
Then you should create more test data to test other scenario (e.g aud as array, non string claims). I also don't like the base64 input; it hides all the information to check the test itself.

Alternatively, I suggest to add this function to the AuthUtils (as private static), and use mock  Json object for test.

Reply:
Done.